ractor_actors/examples/tcp.rs

```diff
@@ -0,0 +1,241 @@
+// LICENSE-MIT file in the root directory of this source tree.
+
+extern crate ractor_actors;
+
+use chrono::{DateTime, Utc};
+use ractor::{Actor, ActorProcessingErr, ActorRef};
+use ractor_actors::net::tcp::listener::*;
+use ractor_actors::net::tcp::session::*;
+use ractor_actors::net::tcp::*;
+use ractor_actors::watchdog;
+use ractor_actors::watchdog::TimeoutStrategy;
+use std::error::Error;
+use std::str::FromStr;
+use std::time::SystemTime;
+
+struct MyServer;
+
+struct MyServerState {}
+struct MyServerArgs {
+    /// Controls if the watchdog is used
+    watchdog: bool,
+
+    /// The port to listen on
+    port: NetworkPort,
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl Actor for MyServer {
+    type Msg = ();
+    type State = MyServerState;
+    type Arguments = MyServerArgs;
+
+    async fn pre_start(
+        &self,
+        myself: ActorRef<Self::Msg>,
+        MyServerArgs { watchdog, port }: Self::Arguments,
+    ) -> Result<Self::State, ActorProcessingErr> {
+        let acceptor = MyServerSocketAcceptor { watchdog };
+
+        let _ = myself
+            .spawn_linked(
+                Some(format!("listener-{}", port)),
+                Listener::new(),
+                ListenerStartupArgs {
+                    port,
+                    encryption: IncomingEncryptionMode::Raw,
+                    acceptor,
+                },
+            )
+            .await?;
+
+        Ok(Self::State {})
+    }
+}
+
+struct MyServerSocketAcceptor {
+    watchdog: bool,
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl SessionAcceptor for MyServerSocketAcceptor {
+    async fn new_session(&self, stream: NetworkStream) -> Result<(), ActorProcessingErr> {
+        tracing::info!("New connection: {}", stream.peer_addr());
+        Actor::spawn(
+            Some(format!("MySession-{}", stream.peer_addr().port())),
+            MySession {
+                info: stream.info(),
+            },
+            MySessionArgs {
+                watchdog: self.watchdog,
+                stream,
+            },
+        )
+        .await
+        .map_err(|e| ActorProcessingErr::from(e))?;
+
+        Ok(())
+    }
+}
+
+struct MySession {
+    info: NetworkStreamInfo,
+}
+
+enum MySessionMsg {
+    FrameReady(Frame),
+}
+
+struct MySessionState {
+    watchdog: bool,
+    session: ActorRef<TcpSessionMessage>,
+}
+
+struct MySessionArgs {
+    watchdog: bool,
+    stream: NetworkStream,
+}
+
+struct MyFrameReceiver {
+    session: ActorRef<MySessionMsg>,
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl FrameReceiver for MyFrameReceiver {
+    async fn frame_ready(&self, f: Frame) -> Result<(), ActorProcessingErr> {
+        self.session
+            .cast(MySessionMsg::FrameReady(f))
+            .map_err(ActorProcessingErr::from)
+    }
+}
+
+#[cfg_attr(feature = "async-trait", async_trait::async_trait)]
+impl Actor for MySession {
```

Comment:
OK I think there's a bit of a misunderstanding in the pattern I intended here.

I believe you shouldn't need this `MySession` actor at all. By defining the frame receiver trait to do the higher-level logic you're doing inside the `handle()` method below, you can get rid of this actor all together.

The `TcpSession` will call your frame receiver caller for you, and you can do whatever conversion/decoding is necessary there.

Reply:
I guess the only reason might be the integration to the watchdog bit here. But I'm not really clear _why_ you're compounding them together. Is it just to check if the session had activity in a certain time period?

I believe you could do this on the frame receiver, and for every frame ping the watchdog, instead of having a custom actor to do it.

Reply:
The watchdog is to ensure client interactivity on a higher level than the TCP timeouts. I wanted to include it in the example as a way to show off how actors are automatically killed (and thus resources cleaned up).

I guess this example might be outdated after my last refactoring, I think I have to take a step back and look at everything again, in particular with the FrameReceiver change idea you mentioned.